File: c++/UNUSED/reader.cpp
/*
	Reader - splits text into words and provides next/previous interface.

	Copyright (c) 2022 Frank McIngvale, see LICENSE
*/
#include "xmalloc.hpp"
#include "reader.hpp"
#include "errors.hpp"
#include <iostream>
#include <sstream>

using namespace std;

Reader::Reader() {
	objlist = new ObjList();
	pos = 0;
}

// add more text to current context
void Reader::addText(const string &text) {
	istringstream input(text);
	string word;
	while(true) {
		input >> word; // read words, delimited by whitespace
		if(input.fail()) {
			return; // reached end of file (probably some whitespace was left, so failed to read a word)
		}
		objlist->push_back(newSymbol(word.c_str(), word.length()));
	}
}

void Reader::clearAll() {
	stack_objlists.clear();
	stack_pos.clear();
	delete objlist;
	objlist = new ObjList();
	pos = 0;
}

// push current context and switch to new context
void Reader::pushObjList(ObjList *objs) {
	stack_objlists.push_back(objlist);
	stack_pos.push_back(pos);
	
	objlist = objs;
	pos = 0;
}

void Reader::popObjList() {
	objlist = stack_objlists.back();
	pos = stack_pos.back();
	stack_objlists.pop_back();
	stack_pos.pop_back();
}

bool Reader::hasPushedObjLists() {
	return stack_objlists.size() > 0;
}

const Object Reader::nextObj() {
	if(pos >= objlist->size()) {
		return NULLOBJ;
	}
	else {
		return objlist->at(pos++);
	}
}

const Object Reader::peekObj() {
	if(pos >= objlist->size()) {
		return NULLOBJ;
	}
	else {
		return objlist->at(pos);
	}
}

const Object Reader::prevObj() {
	if(pos <= 0) {
		return NULLOBJ;
	}
	else {
		return objlist->at(--pos);
	}
}

const Object Reader::peekPrevObj() {
	if(pos <= 0) {
		return NULLOBJ;
	}
	else {
		return objlist->at(pos-1);
	}
}

void Reader::deletePrevObj() {
	if(pos == 0) {
		throw LangError("No previous word to delete!");
	}
	objlist->erase(objlist->begin()+pos-1);
	--pos;
}

void Reader::insertPrevObj(const Object& obj) {
	objlist->insert(objlist->begin()+pos, obj);
	++pos;
}

void Reader::debugPrintObjList() const {
	cout << "Reader:objlist>> ";
	for(auto obj : *objlist) {
		cout << obj.fmtStackPrint() << " ";
	}
	cout << endl;
}

File: lua/UNUSED/reader.lua
--[[
	Reader - splits text into words and provides next/previous interface.

	Copyright (c) 2022 Frank McIngvale, see LICENSE

	This was ported from the Python implementation.
]]

Reader = {}

function Reader:addText(text)
	for word in string.gmatch(text, "[^%s]+") do
		table.insert(self.wordlist, word)
	end
end

function Reader:clearAll()
	self.wordlist = {}
	self.pos = 1
	self.stack = {}
end

function Reader:nextWord()
	if self.pos > #self.wordlist then
		return ""
	else
		self.pos = self.pos + 1
		return self.wordlist[self.pos-1]
	end
end

function Reader:peekWord()
	if self.pos > #self.wordlist then
		return ""
	else
		return self.wordlist[self.pos]
	end
end

function Reader:prevWord()
	if self.pos <= 1 then
		return ""
	else
		self.pos = self.pos - 1
		return self.wordlist[self.pos]
	end
end

function Reader:pushWords(words)
	--print("READER PUSH WORDS:")
	--for i=1,#words do
	--	print(fmtStackPrint(words[i]))
	--end
	table.insert(self.stack, {self.wordlist,self.pos})
	self.wordlist = words
	--self:debug_print_wordlist()
	self.pos = 1
end

function Reader:hasPushedWords(word)
	return #self.stack >= 1
end

function Reader:popWords(words)
	if #self.stack < 1 then
		error(">>>Empty word stack, cannot pop")
	end
	entry = table.remove(self.stack)
	self.wordlist = entry[1]
	self.pos = entry[2]
end

function Reader:deletePrevWord()
	if self.pos <= 1 then
		error(">>>No previous word to delete")
	end
	table.remove(self.wordlist, self.pos-1)
	self.pos = self.pos - 1
end

function Reader:insertPrevWord(word)
	table.insert(self.wordlist, self.pos, word)
	self.pos = self.pos + 1
end

function Reader:debug_print_wordlist()
	print("CURRENT WORDLIST:")
	for i=1,#self.wordlist do
		print(fmtStackPrint(self.wordlist[i]))
	end
end

function Reader:new(obj)
	setmetatable(obj, self)
	self.__index = self
	obj.__class__ = "Reader"
	obj:clearAll() -- set other fields
	return obj
end

function new_Reader()
	return Reader:new({})
end



